docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+   .parquet("path/to/destination/directory")
+   .start()
+   
+// ========== DF with aggregation ==========
+val aggDF = df.groupBy(“device”).count()
+
+// Print updated aggregations to console
+aggDF
+   .writeStream
+   .outputMode("complete")
+   .format("console")
+   .start()
+
+// Have all the aggregates in an in memory table 
+aggDF
+   .writeStream
+   .queryName("aggregates")    // this query name will be the table name
+   .outputMode("complete")
+   .format("memory")
+   .start()
+
+spark.sql("select * from aggregates).show()   // interactively query in-memory table
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+{% highlight java %}
+// ========== DF with no aggregations ==========
+Dataset[Row] noAggDF = deviceDataDf.select("device").where("signal > 10")   
+
+// Print new data to console
+noAggDF
+   .writeStream()
+   .format("console")
+   .start();
+
+// Write new data to Parquet files
+noAggDF
+   .writeStream()
+   .parquet("path/to/destination/directory")
+   .start();
+   
+// ========== DF with aggregation ==========
+Dataset[Row] aggDF = df.groupBy(“device”).count();
+
+// Print updated aggregations to console
+aggDF
+   .writeStream()
+   .outputMode("complete")
+   .format("console")
+   .start();
+
+// Have all the aggregates in an in memory table 
+aggDF
+   .writeStream()
+   .queryName("aggregates")    // this query name will be the table name
+   .outputMode("complete")
+   .format("memory")
+   .start();
+
+spark.sql("select * from aggregates).show()   // interactively query in-memory table
+{% endhighlight %}
+
+</div>
+<div data-lang="python"  markdown="1">
+
+{% highlight python %}
+# ========== DF with no aggregations ==========
+Dataset[Row] noAggDF = deviceDataDf.select("device").where("signal > 10")   
+
+# Print new data to console
+noAggDF\
+   .writeStream()\
+   .format("console")\
+   .start();
+
+# Write new data to Parquet files
+noAggDF\
+   .writeStream()\
+   .parquet("path/to/destination/directory")\
+   .start();
+   
+# ========== DF with aggregation ==========
+Dataset[Row] aggDF = df.groupBy(“device”).count();
+
+# Print updated aggregations to console
+aggDF\
+   .writeStream()\
+   .outputMode("complete")\
+   .format("console")\
+   .start();
+
+# Have all the aggregates in an in memory table. The query name will be the table name
+aggDF\
+   .writeStream()\
+   .queryName("aggregates")\
+   .outputMode("complete")\
+   .format("memory")\
+   .start();
+
+spark.sql("select * from aggregates).show()   # interactively query in-memory table
+{% endhighlight %}
+
+</div>
+</div>
+
+Using Foreach
+The `foreach` operation allows arbitrary operations to be computer on the output data. As of Spark 2.0, this is available only for Scala and Java, not Python. To use this, you will have to implement the interface `ForeachWriter`, which has methods that gets called whenever there is a sequence of rows generated as output after a trigger. It is important note the following
```

Comment:
operations to be computer => operations to be computed
 methods that gets called =>  methods that get called 
is important note the => is important to note the